### PR TITLE
Ship esm for connect-query (in addition to cjs)

### LINKED ISSUE #17: ES Module shipped in a CommonJS package?
Hey! Love the direction of this library and just did a huge refactor to convert all of our grpc calls to using connect query!

I'm in the middle of fixing up our tests and got this message from vistest:
```
Module /Users/greg/code/phc/location-fe/node_modules/@bufbuild/connect-query/dist/src/index.js:14 seems to be an ES Module but shipped in a CommonJS package. You might want to create an issue to the package "@bufbuild/connect-query" asking them to ship the file in .mjs extension or add "type": "module" in their package.json.
```

I'm not sure what the best path forward for you lib is, but I thought I would let you know about the warning.

Vitest also posted a work around, so I think I'm fine for now:
```
As a temporary workaround you can try to inline the package by updating your config:

// vitest.config.js
export default {
  test: {
    deps: {
      inline: [
        "@bufbuild/connect-query"
      ]
    }
  }
}
```

Cheers and thanks for all the work! 🎉

### PR DESCRIPTION
**Pull request by dimitropoulos**
closes #17 

It doesn't seem like this needs to also be done for the protoc generator, but please someone correct me if I'm wrong.

### CODE REVIEW

**Comment by getinnocuous** [comment]
Are there any plans to merge this?

**Comment by sher** [comment]
Would also like to know if there is any progress on this?

**Comment by dimitropoulos** [comment]
@timostamm would you mind taking another look.  I gave this one a completely fresh start, and I think it is hopefully close.  Are there extra fields that should be added to the `/build/cjs/package.json` and `/build/esm/package.json` files?  Maybe `dependencies` related fields?

File: packages/connect-query/package.json
Comment (timostamm) [inline packages/connect-query/package.json:22]: Nit: I'd argue that post-build.ts should be run in the build:esm script.

**Review by timostamm — APPROVED** [review]
You can ship this with `type: module` (just need to swap the package.json in dist) because this doesn't use subpath-exports, and isn't going to run into the dual package hazard. 

But I'm also fine with shipping this as is to unblock users 🙂

File: packages/connect-query/package.json
Comment (dimitropoulos) [reply packages/connect-query/package.json:22]: did you see the [other one](https://github.com/bufbuild/connect-query/blob/57bc26898734056b14193d0b3a96075909fefc18/packages/protoc-gen-connect-query/scripts/post-build.ts#L30)?  Indeed, I did originally do it exactly as you suggest (`post-build-cjs.ts`), but then the other package needed to also make a change to the `esm` so I made both `post-build` in the name of consistency.

whatddya think?

Comment (fubhy) [reply packages/connect-query/package.json:22]: (That's maintained by one of the TS maintainers: https://github.com/arethetypeswrong/arethetypeswrong.github.io)

Comment (timostamm) [reply packages/connect-query/package.json:22]: Hehe, that's awesome @fubhy 😆

Comment (timostamm) [reply packages/connect-query/package.json:22]: The nit was just about moving the call to the npm script @dimitropoulos, so that makes sense on it's own. 

I did not register the two scripts are different. But wait... Your package is CommonJS, so you need to put a `type:module` into dist/esm. 

Would be interesting to see if `arethetypeswrong` registers the package.json.

**Comment by sher** [comment]
@dimitropoulos Thank you, will give alpha a try.

**Comment by paul-sachs** [comment]
@dimitropoulos I tried it out on the marketing site for buf (which is now using connect-query 🎉 ) but am still running into this issue. I have to explicitly tell nextjs to transpile connect-query:

```
error - Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/Users/paulsachs/code/buf.build/node_modules/@bufbuild/connect-query/dist/esm/src/create-query-service' imported from /Users/paulsachs/code/buf.build/node_modules/@bufbuild/connect-query/dist/esm/src/index.js
    at new NodeError (node:internal/errors:372:5)
    at finalizeResolution (node:internal/modules/esm/resolve:437:11)
    at moduleResolve (node:internal/modules/esm/resolve:1009:10)
    at defaultResolve (node:internal/modules/esm/resolve:1218:11)
    at ESMLoader.resolve (node:internal/modules/esm/loader:580:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:294:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:80:40)
    at link (node:internal/modules/esm/module_job:78:36)
```

I'm guessing it's because these files don't include the `.js` extensions.

**Comment by sher** [comment]
@dimitropoulos We are facing the same issue @paul-sachs mentioned
```
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/Users/sher/app/node_modules/@bufbuild/connect-query/dist/esm/src/create-query-service' imported from /Users/sher/app/node_modules/@bufbuild/connect-query/dist/esm/src/index.js
    at new NodeError (node:internal/errors:399:5)
    at finalizeResolution (node:internal/modules/esm/resolve:326:11)
    at moduleResolve (node:internal/modules/esm/resolve:945:10)
    at defaultResolve (node:internal/modules/esm/resolve:1153:11)
    at nextResolve (node:internal/modules/esm/loader:163:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:838:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:77:40)
    at link (node:internal/modules/esm/module_job:76:36) {
  code: 'ERR_MODULE_NOT_FOUND',
  page: '/'
}
```

**Comment by dimitropoulos** [comment]
ok, thanks for the feedback, I'll keep trying.  There's another approach we can take but it creates a problem for the connect-query codebase.  I'll give that a shot.

**Comment by dimitropoulos** [comment]
I think I [fixed it](https://arethetypeswrong.github.io/?p=%40bufbuild%2Fconnect-query%400.2.0-alpha.4) now, maybe? (pushed to a [different branch](https://github.com/bufbuild/connect-query/commits/dimitri/cq-module) until I see if this worked).

**Comment by dimitropoulos** [comment]
nix the above: @paul-sachs and I found a different problem with it in manually testing it.  I am closing this PR and trying (:yawning_face: yet again) a different approach in https://github.com/bufbuild/connect-query/pull/39.

**Pull request closed without merging**
